docs/architecture/adr-105-refactor-mempool-senders.md

```diff
@@ -0,0 +1,141 @@
+    CheckTx(tx types.Tx, cb func(*abci.ResponseCheckTx), txInfo TxInfo) error
+    ```
+    to
+    ``` golang
+    CheckTx(tx types.Tx) (abcicli.ReqRes, error)
+    ```
+  - The returning `ReqRes` object can be used to set and invoke a callback to
+    handle the response, if needed.
+  - The callback parameter `cb` is no longer needed. Currently, this is mainly
+    used by the RPC endpoints `broadcast_tx_sync` and `broadcast_tx_commit`, and
+    in tests for checking that the response is valid. However, the same
+    functionality can be obtained with `ReqRes` response.
+  - `txInfo` contains information about the sender, which is also no longer
+    needed, as justified by the next point.
+- The list of senders for each transaction is currently stored in `mempoolTx`,
+  the data structure for the entries of `txs`. Move the senders out of
+  `mempoolTx` to a new map `txSenders` of type `map[types.TxKey]map[uint16]bool`
+  in the mempool reactor. `txSenders` would map transaction keys to a set of
+  peer ids (of type `uint16`). Add also a `cmtsync.RWMutex` lock to handle
+  concurrent accesses to the map.
+  - This refactoring should not change the fact that the list of senders live as
+    long as the transactions are in the mempool. When a transaction is received
+    by the reactor (either via RPC or P2P), we call `CheckTx`. We know whether a
+    transaction is valid and was included in the mempool by reading the `ReqRes`
+    response. If this is the case, record the list of senders in `txSenders`.
+    When a transaction is removed from the mempool, notify the reactor to remove
+    the list of senders for that transaction, with the channel described below.
+- In `CListMempool`, introduce a new channel `txsRemoved` of type `chan
+  types.TxKey` to notify the mempool reactor that a transaction was removed from
+  the mempool.
+  - This is the same mechanism used to notify the consensus reactor when there
+  are transactions available to include in a block.
+- In the mempool reactor, introduce a new goroutine to handle incoming
+  transaction keys from the `txsRemoved` channel. For each key received, update
+  `txSenders`.
+- In `CListMempool`, `resCbFirstTime` is the function that handles responses of
+  type `CheckTxType_New`. Instead of setting it as an ad-hoc callback on each
+  transaction, we could now call it directly from `globalCb`, where responses of
+  type `CheckTxType_Recheck` are already being handled.
+
+## Consequences
+
+The refactoring proposed here does not affect how users and other peers
+interact with the mempool. It will only change how transaction metadata is
+stored internally.
+
+### Positive
+
+- Get rid of the complex design pattern of callbacks for handling
+  `CheckTxType_New` responses.
+- Clear separation of propagation protocol in the reactor and the mempool
+  implementation.
+- Allow for future improvements to both the propagation protocol and the mempool
+  implementation.
+
+### Negative
+
+None
```

Comment:
How confident are we that we are well covered by the current set of unit tests that are testing this code?
IOW, do we have a rough idea of:
* how costly it would be to fix the existing ones
* how many new ones we would have to write
* how likely it is that the (adapted) old ones and the new ones will catch problems in the refactoring

Reply:
In the draft PR (#1010), I already fixed the existing tests. There were many changes, but not difficult. We'll certainly have to extend the test scenarios, at least to cover the new data structure `txSenders`, and probably a bit more. I still have to check what the tests are not covering.

In any case, I think all these are implementation details, not related to the architecture or design. What we could do is to add to the Neutral section something like "We need to extend the existing tests to cover new scenarios related to the new data structures and some potential concurrent issues".

Reply:
💯